simplify IKFast tutorial (#332)
ros-planning/moveit_tutorials: doc/ikfast/ikfast_tutorial.rst

IKFast Kinematics Solver
========================

> 

Author: The image is not expressive for this tutorial.

> 

Reviewer: During our last refactor, we had a goal to add images to every tutorial to make them feel more friendly and less boring.

This picture @mlautman made using OpenRave, which is very relevant to this tutorial. please restore it.

> 

Author: I know that this image is from OpenRave. However, I'm not convinced yet that it is useful in any fashion for this tutorial.
For me, it's just distracting. Following the given tutorial, the user will never see a similar image.

> 

Reviewer: This isn't a technical question, its a user-interface question. Many users of the tutorials might be intimidated by the complexity of moveit, pictures soften that blow in some kind of psychological way. If you have a better pic, please suggest. Otherwise, let's keep this work @mlautman put into the tutorials

> 

Author: I bow to the majority and will restore the image. However, I'm still not convinced.

In this section, we will walk through configuring an IKFast plugin for MoveIt.

What is IKFast?
---------------

IKFast, the Robot Kinematics Compiler, is a powerful inverse kinematics solver provided within Rosen Diankov's `OpenRAVE <http://openrave.org>`_ motion planning software. IKFast automatically analyses any complex kinematic chain for common patterns that allow for an analytic solution and generates C++ code to find them.
As a consequence, IKFast provides extremely stable solutions that can be found in a few microseconds on recent processors.

MoveIt IKFast
---------------

MoveIt provides tools to generate an IKFast kinematics plugin for MoveIt using the OpenRAVE generated cpp files.
This tutorial will step you through setting up your robot to utilize the power of IKFast.
MoveIt IKFast is tested on ROS Melodic with a 6DOF and 7DOF robot arm manipulator.

> While it works in theory, MoveIt IKFast doesn't support >7 degree of freedom arms.

Reviewer: ```suggestion
While it works in theory, MoveIt IKFast doesn't currently support >7 degree of freedom arms.
```

Getting Started
-----------------
If you haven't already done so, make sure you've completed the steps in `Getting Started <../getting_started/getting_started.html>`_.

You should have a MoveIt configuration package for your robot that was created by using the `Setup Assistant <../setup_assistant/setup_assistant_tutorial.html>`_.

OpenRAVE is a planning framework as complex as MoveIt itself and installing it is tricky -- particularly because it is not actively developed anymore.

> Fortunately, personalrobotics provide a `docker image <https://hub.docker.com/r/personalrobotics/ros-openrave>`_ based on Ubuntu 14.04 with OpenRAVE 0.9.0 and ROS Indigo installed, which can be used to generate the solver code once.

Reviewer: this is a nice approach

So the easiest way to run the IKFast code generator is through this docker image.

> For manual building instructions (tailored towards Ubuntu 16.04), please see the `Kinetic version of this tutorial <http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/ikfast/ikfast_tutorial.html>`_.

Author: I don't see a conflict in this. If people insist on using the tedious, manual way, the can do so by following the old tutorial.

To follow the recommended, docker-based approach, ensure you have docker installed and started: ::

 sudo apt-get install docker.io
 sudo service docker start

Ensure that you can run docker with your user account: ::

>  sudo usermod -a -G docker $USER

Author: There is no output to verify. This command just adds `$USER` to the docker group, thus enabling `$USER` to run docker.

>  sudo usermod -a -G docker $USER

Author: This is weird to me. 
If there is an error with the command, I guess an appropriate error message is written.

>  sudo usermod -a -G docker $USER

Author: Looks like you looked at an old version of the PR. Not sure how this is possible at all...

>  sudo usermod -a -G docker $USER

Reviewer: i think this is good as @rhaschke has it

  export MYROBOT_NAME="panda"

OpenRAVE uses Collada instead of URDF to describe the robot. In order to automatically convert your robot's URDF to Collada, you need to provide the .urdf file.
If your .urdf file is generated from `xacro <http://wiki.ros.org/xacro/>`_ files you can generate the urdf using the following command: ::

  rosrun xacro xacro -o $MYROBOT_NAME.urdf $MYROBOT_NAME.urdf.xacro

Select IK Type
^^^^^^^^^^^^^^
You need to choose which type of IK you want to solve for. See `this page <http://openrave.org/docs/latest_stable/openravepy/ikfast/#ik-types>`_ for more info.
The most common IK type is *transform6d*.

Choose Planning Group
^^^^^^^^^^^^^^^^^^^^^
If your robot has more than one arm or "planning group" that you want to generate an IKFast solver for, you need to repeat the following process for each group.
The following instructions will assume you have chosen one ``<planning_group_name>``. Furthermore, you need to know the names of the base link and the endeffector link of the chain to solve for.

Generate IKFast MoveIt plugin
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To generate the IKFast MoveIt plugin, issue the following command: ::

>   rosrun moveit_kinematics auto_create_ikfast_moveit_plugin.sh --iktype Transform6D $MYROBOT_NAME.urdf <planning_group_name> <base_link> <eef_link>

Reviewer 2: This currently fails on my machine with:

```
WARNING: The following packages cannot be authenticated!
  collada-dom2.4-dp-base collada-dom2.4-dp-dev collada-dom-dev
  ros-indigo-angles ros-indigo-urdf-parser-plugin ros-indigo-collada-parser
  ros-indigo-eigen-stl-containers ros-indigo-octomap ros-indigo-random-numbers
  ros-indigo-resource-retriever ros-indigo-geometric-shapes
  ros-indigo-tf2-msgs ros-indigo-tf2 ros-indigo-tf2-py ros-indigo-tf2-ros
  ros-indigo-tf ros-indigo-urdf ros-indigo-collada-urdf
E: There are problems and -y was used without --force-yes
```

Reviewer 2: Also, this does not sufficiently cover the command line arguments for  `auto_create_ikfast_moveit_plugin.sh`

Reviewer 2: I spent considerable time making sure that the reader could follow along with this tutorial line copy-pasting in commands for their own robot. I would like to maintain that usability.

Author: The failure is due to the ROS key change. I pushed a corresponding fix to master + melodic-devel as this also blocked Travis (https://github.com/ros-planning/moveit/pull/1503).

Reviewer 2: This leaves out many of the command line arguments used by `auto_create_ikfast_moveit_plugin.sh` Please add a note that `--help` will provide documentation for the command line arguments.

Reviewer 2: I opened an issue about the gpg key issue: https://github.com/personalrobotics/docker-public-images/issues/4

Reviewer 2: I think we will have to wait until the docker is re-built before we can proceed with this PR

Author: No, not at all. We have a hot-fix for the GPG key issue via https://github.com/ros-planning/moveit/pull/1503.

Reviewer 3: When I run this it fails. As a note $MYROBOT_NAME is set to panda_arm_hand as per the urdf I was using.
```
[ERROR] [1563206287.111715886]: Could not open file [robot.urdf] for parsing.
[ERROR] [1563206287.111754005]: failed to open urdf file robot.urdf
```
I also think that a quick explanation of where to use this command and possibly an example for the default panda arm config made 
```
rosrun moveit_kinematics auto_create_ikfast_moveit_plugin.sh --iktype Transform6D $MYROBOT_NAME.urdf panda_arm panda_link0 panda_link8
```

  rosed "$MYROBOT_NAME"_moveit_config kinematics.yaml

Edit these parts: ::

 <planning_group>:
   kinematics_solver: <myrobot_name>_<planning_group>_ikfast_plugin/IKFastKinematicsPlugin

Test the Plugin
^^^^^^^^^^^^^^^
Use the MoveIt RViz Motion Planning Plugin and use the interactive markers to see if correct IK Solutions are found.

Updating the Plugin
-------------------

If any future changes occur with MoveIt! or IKFast, you might need to re-generate this plugin using our scripts. To facilitate this, a bash script was automatically created in the root of your IKFast MoveIt package, named *update_ikfast_plugin.sh*. This regenerates the plugin from the OpenRAVE-generated .cpp solver file.

Tweaking the creation process
-----------------------------

The process of creating the IKFast MoveIt plugin comprises several steps:

> 1. Download (and slightly modify) the docker image provided by `personalrobotics <https://hub.docker.com/r/personalrobotics/ros-openrave>`_

Author: Why should these modifications be documented here? They are implementation details.

> 1. Download (and slightly modify) the docker image provided by `personalrobotics <https://hub.docker.com/r/personalrobotics/ros-openrave>`_

Reviewer 4: I guess what @mlautman is implying is that from the language it sounds like the reader/user would have to "slightly modify" the image -- without getting any additional information on how to do that.

> 1. Download (and slightly modify) the docker image provided by `personalrobotics <https://hub.docker.com/r/personalrobotics/ros-openrave>`_

Author: Thanks for this hint. I will improve the language accordingly.

2. Convert the ROS URDF file to Collada required for OpenRAVE: ::

>      rosrun collada_urdf urdf_to_collada <myrobot_name>.urdf <myrobot_name>.dae

Reviewer 2: use $MYROBOT_NAME

Author: Yep. This one I forgot to change. Fixed.

   Sometimes floating point issues arise in converting a URDF file to Collada, which prevents OpenRAVE to find IK solutions.
   Using a utility script, one can easily round all numbers down to n decimal places in your .dae file.
   If the generator takes longer than, say an hour, try the following: ::

>      rosrun moveit_ikfast round_collada_numbers.py <input_dae> <output_dae> <decimal places>

Author: For the yaml code snippet you didn't used the environment variable. So did I.

   From experience we recommend 5 decimal places, but if the OpenRave ikfast generator takes to long to find a solution, lowering the number of decimal places should help. For example: ::

>      rosrun moveit_ikfast round_collada_numbers.py <myrobot_name>.dae <myrobot_name>.rounded.dae 5

Author: OK.

> 3. Run OpenRAVE IKFast tool to generate C++ solver code

Author: Obviously it didn't became clear that these steps are _automatically_ performed by the `auto*.sh` script. There is no need for the user to explicitly run OpenRave. Hopefully the new text resolves this misunderstanding. If not, please feel free to edit the text directly.

4. Create the MoveIt IKFast plugin package wrapping the generated solver

> To manually tweak the process, you can start the script ``auto_create_ikfast_moveit_plugin.sh`` from any intermediate step by providing the corresponding output from the previous step as input (``.dae`` or ``.cpp``) instead of the ``.urdf`` file.

Reviewer 2: ```suggestion
The ``auto_create_ikfast_moveit_plugin.sh`` script reads in the file extension from the input to determine which steps to run. To re-run the script from any intermediate step, simply provide the corresponding output from the previous step as input (``.dae`` or ``.cpp``) instead of the ``.urdf`` file.
```

Author: OK. Slightly modified.

 rosdep install -y --from-paths . --ignore-src --rosdistro ${ROS_DISTRO}
 catkin build

Creating the IKFast MoveIt plugin
---------------------------------

To facilitate copy-and-paste, we suggest to define the robot name as an environment variable: ::

>   export MYROBOT_NAME="panda"

Reviewer: ```suggestion
  export MYROBOT_NAME="panda_arm"
```
@simonGoldstein this is what you mean, right? +1

OpenRAVE uses Collada instead of URDF to describe the robot. In order to automatically convert your robot's URDF to Collada, you need to provide the .urdf file.
If your .urdf file is generated from `xacro <http://wiki.ros.org/xacro/>`_ files you can generate the urdf using the following command: ::

> 

Reviewer: @rhaschke did you address this issue somehow? i think we should keep this work around I made, many years ago...

>   rosrun xacro xacro -o $MYROBOT_NAME.urdf $MYROBOT_NAME.urdf.xacro

Author: This needs to be run in `catkin_ws/src`, because there the new package will be created (in the current working dir).
If your .urdf file resides somewhere else, you need to provide the full path. For that reason, I'm not a fan of using the `MYROBOT_NAME` variable...

Select IK Type
^^^^^^^^^^^^^^
You need to choose which type of IK you want to solve for. See `this page <http://openrave.org/docs/latest_stable/openravepy/ikfast/#ik-types>`_ for more info.
The most common IK type is *transform6d*.

Choose Planning Group
^^^^^^^^^^^^^^^^^^^^^
If your robot has more than one arm or "planning group" that you want to generate an IKFast solver for, you need to repeat the following process for each group.
The following instructions will assume you have chosen one ``<planning_group_name>``. Furthermore, you need to know the names of the base link and the endeffector link of the chain to solve for.

Generate IKFast MoveIt plugin
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To generate the IKFast MoveIt plugin, issue the following command: ::

> 

Reviewer 2: If we can preserve the existing method while also supporting the auto ikfast script, then we can disregard this comment

> 

Author: This section is not needed with the new script anymore. You directly specify the base and eef link _names_ and these should be clear to any user.

> 

Reviewer 4: It may be 'prudent' to explain/keep around the link id part. For unknown reasons the ikfast generators can sometimes get confused when using names, and link ids have saved me once or twice in those situations.

> 

Author: @gavanderhoorn, Do you want to add a proposal for this paragraph?

> 

Reviewer 4: Tbh I'm really overloaded right now, but just wanted to relay my experiences so they could potentially be taken into account.

sorry :(

> 

Reviewer: if we just keep the old explanation of how to output the link names, this should be resolved.

  rosrun moveit_kinematics auto_create_ikfast_moveit_plugin.sh --iktype Transform6D $MYROBOT_NAME.urdf <planning_group_name> <base_link> <eef_link>

The speed and success of this process will depend on the complexity of your robot. A typical 6 DOF manipulator with 3 intersecting axes at the base or wrist will take only a few minutes to generate the solver code. For a detailed explanation of the creation procedure and additional tweaks of the process, see `Tweaking the creation process`_.

> The above command creates a new ROS package named ``$MYROBOT_NAME_<planning_group_name>_ikfast_plugin`` within the current folder.

Reviewer 3: "The above command" sounds awkward. It may be better to replace it with either "The command above" or "The previous command"

Thus, you need to rebuild your workspace so the new package is detected: ::

  catkin build

Usage
-----
The IKFast plugin can be used as a drop-in replacement for the default KDL IK Solver, but with greatly increased performance. The MoveIt configuration file should be automatically edited by the generator script but in some cases this might fail. In this situation you can switch between the KDL and IKFast solvers using the *kinematics_solver* parameter in the robot's kinematics.yaml file: ::

>   rosed "$MYROBOT_NAME"_moveit_config kinematics.yaml

Reviewer 3: Unless I am missing something I don't think this needs the space.

Author: It does in order to be formatted as a command.

Edit these parts: ::

 <planning_group>:
   kinematics_solver: <myrobot_name>_<planning_group>_ikfast_plugin/IKFastKinematicsPlugin

Test the Plugin
^^^^^^^^^^^^^^^

> Use the MoveIt RViz Motion Planning Plugin and use the interactive markers to see if correct IK Solutions are found.

Reviewer 3: I think it would be really nice to add more information on how to actually test this with RViz.

Author: Sorry, but using rviz interactive markers was handled in another tutorial already.

Updating the Plugin
-------------------

If any future changes occur with MoveIt! or IKFast, you might need to re-generate this plugin using our scripts. To facilitate this, a bash script was automatically created in the root of your IKFast MoveIt package, named *update_ikfast_plugin.sh*. This regenerates the plugin from the OpenRAVE-generated .cpp solver file.

Tweaking the creation process
-----------------------------

> The process of creating the IKFast MoveIt plugin comprises several steps, performed one by one by the creation script:

Reviewer 2: ```suggestion
The process of creating the IKFast MoveIt plugin comprises several steps, performed one-by-one by the creation script:
```
